server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -249,6 +249,47 @@ public List<Pair<DataSegment, String>> retrieveUsedSegmentsAndCreatedDates(Strin
     );
   }
 
+  @Override
+  public List<DataSegment> retrieveUnusedSegmentsForExactIntervalAndVersion(
+      String dataSource,
+      Interval interval,
+      String version
+  )
+  {
+    StringBuilder sb = new StringBuilder();
+    sb.append(StringUtils.format("SELECT payload FROM %s", dbTables.getSegmentsTable()));
```

Comment:
For the use case at hand, we don't even need to get the payload, we can just do `SELECT MAX(id)`. Then you can get rid of the `for` loop in `getUnusedMaxId` method.

Reply:
segment_111 is less than segment_2 lexicographically. I don't think this would work

Reply:
Good point. But we should still avoid fetching the payload and deserializing it when all we need is the `id`.

Reply:
Is there an existing way to parse the partition number from the id?

Reply:
Yeah, parse the string id as a `SegmentId` object.

Reply:
Done, thanks